### PR TITLE
log: skip on unknown nested mixing instead of stopping the logging

### LINKED ISSUE #874: logging: Exception logging might stop in the middle of a nested_exception
If the exception logging function encounters an `std::nested_exception` that is not mixed in with an `std::exception` or it's
derivatives, the logging will stop and will not continue to parse the rest of the nested_exception chain with might be known.
Moreover, there is no indication in the logging that this might have happened even though it means that there is a possibility
there is an ill formed code that abuse the nested_exception or throwing system.

### PR DESCRIPTION
**Pull request by eliransin**
The logging function for an exception type will stop if a nested
exception change is not mixed in with a type derived from
std::exception. The consequence of this is that if there are more
nested exceptions in the chain they will not be printed out which
means that we can lose viable debugging information.
This commit fixes this by adding a catch handler that will skip the
unknown part of the chain and will try to continue to the next part of
the exception chain which is the std::nested_exception member.

It also adds explicit indication in the logging where there is no
information on an object - for example when it is not derived from
std::exception.

Tests for the exception logging are added for all of the different
cases.

Fixes #874

### CODE REVIEW

File: src/util/log.cc [hunk 1]
```diff
@@ -553,6 +553,23 @@ std::ostream& operator<<(std::ostream& out, const std::exception_ptr& eptr) {
             } catch (...) {
                 out << ": " << std::current_exception();
             }
+        } catch (const std::nested_exception& ne) {
+           // if we got here, it means that the nested_exception is not mixed in
+           // with an object of type std::exception this it can mean one of two
```
Comment (tgrabiec) [inline src/util/log.cc:558]: ` this it` -> `. It`

File: src/util/log.cc [hunk 2]
```diff
@@ -553,6 +553,23 @@ std::ostream& operator<<(std::ostream& out, const std::exception_ptr& eptr) {
...
+           // things:
+           // 1. The std::nested_exception was built and thrown using a normal
+           //    throw and not in the context of another catch with std::throw_with_nested
+           // 2. Somewhere in the code there is a throw of a type which doesn't derive from
+           //    std::exception and then it is rethrown using std::throw_with_nested.
+           out<< ": " << "unknown nested type";
```
Comment (tgrabiec) [inline src/util/log.cc:564]: Missing space after `out`.

File: src/util/log.cc [hunk 3]
```diff
@@ -553,6 +553,23 @@ std::ostream& operator<<(std::ostream& out, const std::exception_ptr& eptr) {
...
+           if (ne.nested_ptr()) {
+                // skip over the mixed in object and continue recursively to the
+                // unfold the contained exception itself
+                out <<  ": " << ne.nested_ptr();
```
Comment (tgrabiec) [inline src/util/log.cc:568]: Double space after first <<

File: src/util/log.cc [hunk 4]
```diff
@@ -553,6 +553,23 @@ std::ostream& operator<<(std::ostream& out, const std::exception_ptr& eptr) {
...
+           } else {
+               // there is no contained exception
+               out<< ": " << "empty contained exception";
```
Comment (tgrabiec) [inline src/util/log.cc:571]: Missing space after `out`.

**Review by tgrabiec — CHANGES_REQUESTED** [review]
I only have some cosmetic issues

**Review by tgrabiec — APPROVED** [review]

File: src/util/log.cc [hunk 5]
```diff
@@ -553,6 +553,23 @@ std::ostream& operator<<(std::ostream& out, const std::exception_ptr& eptr) {
...
             }
+        } catch (const std::nested_exception& ne) {
+           // if we got here, it means that the nested_exception is not mixed in
+           // with an object of type std::exception. It can happen in one of two
+           // cases:
+           // 1. The std::nested_exception was built and thrown using a normal
+           //    throw and not in the context of another catch with std::throw_with_nested
+           // 2. Somewhere in the code there is a throw of a type which doesn't derive from
+           //    std::exception and then it is rethrown using std::throw_with_nested.
+           out << ": " << "unknown nested type";
```
Comment (nyh) [inline src/util/log.cc:564]: Why is this an "unknown nested type"?  When we print ne.nested_ptr() below, doesn't it know it's own type? We have std::ostream& operator<<(std::ostream& out, const std::exception_ptr& eptr) which does exactly that.

**Review by nyh — CHANGES_REQUESTED** [review]
I have one question on the "unknown nested type" message below, but what bothers me even more is the commit message. The whole point of this patch is to improve the way we print nested exceptions, so why does the example in the commit message use a different print_exception function? It should use "<<" and show how the code below really prints nested exceptions! This really confused me, when initially I didn't understand how come the output looks different from what the code outputs.

Also I didn't understand why this code is hard to test... Looks very easy to test! Just throw the exceptions exactly like in the commit message, "<<" the resulting exception into an std::stringstream, and compare the string to what you expect. Sounds easy to me, and better than putting code in the commit message (which, as I said above, doesn't even use the operator<< you added).

File: src/util/log.cc (hunk 5 above)
Comment (eliransin) [reply src/util/log.cc:564]: The resulting `std::nested_exception` type when generated properly with `std::throw_with_nested(T object)` is not a pure `std::nested_exception` but a **mix of two types** - a nested_type that *contains* the current exception, and the type
contained in the object. something like `class x : public T, public nested_exception {...` 
We can't account for every T in our code so we also account for objects derived from `std::exception` and then we print 
the what.
The "unknown nested type" print, indicates that we don't know the **whole** nested type, or more specifically, what T is and hence we can't print it.
About the test - fair enough, I will add a test instead the code in the commit message.

Comment (nyh) [reply src/util/log.cc:564]: Hmm, I think I see now what you meant about the "unknown nested type". Ok.

**Comment by eliransin** [comment]
@nyh thanks for insisting on the tests, made me polish the logging code a little bit more :slightly_smiling_face: 
Changes from previous version:
1. Removed some redundancy - the logging was already able to handle empty pointers so no need to handle separately,
moreover it will make a more consistent log.
2. Add an explicit logging of "no more information" so it is easy to deduce the exact flow of this function given a log line.
It is an edge case anyway so if we see this printout we should probably debug the exception throwing code anyway.
3. Added tests to cover all of the cases, old and new.

File: src/util/log.cc [hunk 6]
```diff
@@ -553,8 +553,17 @@ std::ostream& operator<<(std::ostream& out, const std::exception_ptr& eptr) {
...
+            // with an object of type std::exception. It can happen in one of two
+            // cases:
+            // 1. The std::nested_exception was built and thrown using a normal
+            //    throw and not in the context of another catch with std::throw_with_nested
+            // 2. Somewhere in the code there is a throw of a type which doesn't derive from
+            //    std::exception and then it is rethrown using std::throw_with_nested.
+            out << ": " << "unknown nested type: " << ne.nested_ptr();
         } catch(...) {
-            // no extra info
+            out << " (" << "no extra info" << ")";
```
Comment (nyh) [inline src/util/log.cc:566]: Nitpick: this new "no extra info" message is unrelated to the issue of nested exceptions (you have an example in your test below), and I wonder if it's helpful... Basically the operator<< already printed the exception's type, but indeed doesn't know how to print any more information about its value. But I don't have strong feelings either way, so I think this is ok.

**Review by nyh — APPROVED** [review]

File: tests/unit/exception_logging_test.cc [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+            throw unknown_obj(format("Exception Level {}", nesting_level));
+        } else {
+            throw std::runtime_error(format("Exception Level {}", nesting_level));
+        }
+    }
+}
+
+// This function generates the expected logging output string of an exception
+// thrown by the  exception generator function with a specific output.
+std::string exception_generator_str(uint32_t test_instance,int nesting_level) {
```
[the remark this answers is not in the document]
Comment (eliransin) [reply tests/unit/exception_logging_test.cc:73]: I understand that point, for completeness, here are the strings that are produces by the run of this generator test:
```
unknown_obj (no extra info)
std::runtime_error (Exception Level 0)
std::_Nested_exception<unknown_obj>: unknown nested type: unknown_obj (no extra info)
std::_Nested_exception<std::runtime_error> (Exception Level 1): unknown_obj (no extra info)
std::_Nested_exception<unknown_obj>: unknown nested type: std::runtime_error (Exception Level 0)
std::_Nested_exception<std::runtime_error> (Exception Level 1): std::runtime_error (Exception Level 0)
std::_Nested_exception<unknown_obj>: unknown nested type: std::_Nested_exception<unknown_obj>: unknown nested type: unknown_obj (no extra info)
std::_Nested_exception<std::runtime_error> (Exception Level 2): std::_Nested_exception<unknown_obj>: unknown nested type: unknown_obj (no extra info)
std::_Nested_exception<unknown_obj>: unknown nested type: std::_Nested_exception<std::runtime_error> (Exception Level 1): unknown_obj (no extra info)
std::_Nested_exception<std::runtime_error> (Exception Level 2): std::_Nested_exception<std::runtime_error> (Exception Level 1): unknown_obj (no extra info)
std::_Nested_exception<unknown_obj>: unknown nested type: std::_Nested_exception<unknown_obj>: unknown nested type: std::runtime_error (Exception Level 0)
std::_Nested_exception<std::runtime_error> (Exception Level 2): std::_Nested_exception<unknown_obj>: unknown nested type: std::runtime_error (Exception Level 0)
std::_Nested_exception<unknown_obj>: unknown nested type: std::_Nested_exception<std::runtime_error> (Exception Level 1): std::runtime_error (Exception Level 0)
std::_Nested_exception<std::runtime_error> (Exception Level 2): std::_Nested_exception<std::runtime_error> (Exception Level 1): std::runtime_error (Exception Level 0)
```

[the remark this answers is not in the document]
Comment (nyh) [reply tests/unit/exception_logging_test.cc:73]: Thanks. I already committed this series, so if we want to improve it will need to be a followup patch.
In these examples you can see what I asked about the string "unknown exception type"... Check out: 
```
std::_Nested_exception<unknown_obj>: unknown nested type: std::runtime_error (Exception Level 0)
```
What does "unknown nested type" refer to? The message said it's std::nested_exception<unknown_obj> so we know it's an unknown_obj, and, as the next part of the message says - also an std::runtime_error. So what has "unknown type"?

File: src/util/log.cc (hunk 6 above)
Comment (eliransin) [reply src/util/log.cc:566]: I think this helps for completeness of the log message format. In the examples it can be seen the it makes the message more regular or uniform, using an example from https://github.com/scylladb/seastar/pull/875#discussion_r587290842 :
```
unknown_obj (no extra info)
std::runtime_error (Exception Level 0)
```
If anything, I would have added this `(no extra info)` also to the `unknown nested type` remark because it stands out.
On the other hand we should never see either of them if our exception throwing code is well written so I am not sure we
should devote a fix for that.
This is my opinion of course and it is totally subjective :slightly_smiling_face:

File: tests/unit/exception_logging_test.cc (hunk 7 above)
[the remark this answers is not in the document]
Comment (eliransin) [reply tests/unit/exception_logging_test.cc:73]: Now I see the point, maybe I should change it to `non exception derived type (no extra info)`.
I can also just skip this entirely.
I am inclining towards that it is not that important but I could add another patch if it will bother someone.

[the remark this answers is not in the document]
Comment (nyh) [reply tests/unit/exception_logging_test.cc:73]: If you really want that "no extra info" on unknown exception types, you should - I think - at least be consistent and add it here like in non-nested exceptions. I.e., the above message will become:
```
std::_Nested_exception<unknown_obj> (no extra info): std::runtime_error (Exception Level 0)
```
i.e., no "unknown type" message (which feels wrong to me, the type is known), no "non exception derived type", just a "no extra info" exactly like in the other catch(...).

I still don't know what this "no extra info" message adds or clarifies, though. But at least it's not misleading like (I think) "unknown type" is misleading.
